File: src/ios/FirebaseAnalyticsPlugin.m
#import "FirebaseAnalyticsPlugin.h"
#import "OutSystems-Swift.h"

@import AppTrackingTransparency;
@import FirebaseAnalytics;
@import FirebaseCore;

@interface FirebaseAnalyticsPlugin ()

@property (strong, nonatomic) id<OSFANLManageable> manager;

@end

@implementation FirebaseAnalyticsPlugin

- (void)pluginInitialize {
    NSLog(@"Starting Firebase Analytics plugin");

    if(![FIRApp defaultApp]) {
        [FIRApp configure];
    }
    
    self.manager = [OSFANLManagerFactory createManager];
}

- (void)logEvent:(CDVInvokedUrlCommand *)command {
    NSString* name = [command.arguments objectAtIndex:0];
    NSDictionary* parameters = [command.arguments objectAtIndex:1];

    [FIRAnalytics logEventWithName:name parameters:parameters];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)logECommerceEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *argumentsDictionary = [command argumentAtIndex:0];
    NSError *error;
    
    OSFANLOutputModel *outputModel = [self.manager createEventModelFor:argumentsDictionary error:&error];
    if (!outputModel && error) {
        [self sendError:error forCallbackId:command.callbackId];
        return;
    }
    
    [FIRAnalytics logEventWithName:outputModel.name parameters:outputModel.parameters];
    [self sendSuccessfulResultforCallbackId:command.callbackId];
}

- (void)setUserId:(CDVInvokedUrlCommand *)command {
    NSString* id = [command.arguments objectAtIndex:0];

    [FIRAnalytics setUserID:id];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setUserProperty:(CDVInvokedUrlCommand *)command {
    NSString* name = [command.arguments objectAtIndex:0];
    NSString* value = [command.arguments objectAtIndex:1];

    [FIRAnalytics setUserPropertyString:value forName:name];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setEnabled:(CDVInvokedUrlCommand *)command {
    bool enabled = [[command.arguments objectAtIndex:0] boolValue];

    [FIRAnalytics setAnalyticsCollectionEnabled:enabled];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setCurrentScreen:(CDVInvokedUrlCommand *)command {
    NSString* screenName = [command.arguments objectAtIndex:0];

    [FIRAnalytics logEventWithName:kFIREventScreenView parameters:@{
        kFIRParameterScreenName: screenName
    }];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)resetAnalyticsData:(CDVInvokedUrlCommand *)command {
    [FIRAnalytics resetAnalyticsData];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setDefaultEventParameters:(CDVInvokedUrlCommand *)command {
    NSDictionary* params = [command.arguments objectAtIndex:0];

    [FIRAnalytics setDefaultEventParameters:params];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)requestTrackingAuthorization:(CDVInvokedUrlCommand *)command {
    
    bool showInformation = [[command.arguments objectAtIndex:0] boolValue];

    if(showInformation) {
        
        NSString* title = [command.arguments objectAtIndex:1];
        NSString* message = [command.arguments objectAtIndex:2];
        NSString* buttonTitle = [command.arguments objectAtIndex:3];
        
        [self showPermissionInformationPopup:title :message :buttonTitle :^(UIAlertAction *action ) {
            [self showTrackingAuthorizationPopup:command];
        }];
        
    }
    else {
        [self showTrackingAuthorizationPopup:command];
    }
}

- (void)showTrackingAuthorizationPopup:(CDVInvokedUrlCommand *)command {
    
    if (@available(iOS 14, *)) {
        
        NSDictionary *dict = NSBundle.mainBundle.infoDictionary;
        
        if([dict objectForKey:@"NSUserTrackingUsageDescription"]){
            [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
                BOOL result = status == ATTrackingManagerAuthorizationStatusAuthorized;
                
                CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:result];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }];
            return;
        }
    }
    
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:true];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setConsent:(CDVInvokedUrlCommand*)command
{
    NSString *jsonString = [command.arguments objectAtIndex:0];
    
    NSData *jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    NSError *error = nil;
    NSArray *array = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&error];
    
    if (error != nil) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:error.description];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    
    NSMutableDictionary* firebaseConsentDict = [NSMutableDictionary dictionary];
    
    for (NSDictionary *item in array) {
        NSNumber *type = item[@"Type"];
        NSNumber *status = item[@"Status"];
        FIRConsentType consentType = [self consentTypeFromNumber:type];
        FIRConsentStatus consentStatus = [self consentStatusFromNumber:status];
        
        if (consentType && consentStatus) {
            firebaseConsentDict[consentType] = consentStatus;
        } else {
            NSLog(@"Warning: Ignoring invalid consent type or status for: %@ & %@", type, status);
        }
    }
    
    if (firebaseConsentDict.count > 0) {
        [FIRAnalytics setConsent:firebaseConsentDict];
        printf("Success setConsent");
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    } else {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                                        messageAsString:@"No valid consent types provided"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

typedef void (^showPermissionInformationPopupHandler)(UIAlertAction*);
- (void)showPermissionInformationPopup:
(NSString *)title :
(NSString *)message :
(NSString *)buttonTitle :
(showPermissionInformationPopupHandler)confirmationHandler
{
    
    UIAlertController *alert = [UIAlertController
                                alertControllerWithTitle:title
                                message:message
                                preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *okAction = [UIAlertAction
                               actionWithTitle:buttonTitle
                               style:UIAlertActionStyleDefault
                               handler:confirmationHandler];
    
    [alert addAction:okAction];
    [self.viewController presentViewController:alert animated:YES completion:nil];
}

#pragma mark - Result Callback Methods (used for `LogECommerceEvent`)

- (void)sendSuccessfulResultforCallbackId:(NSString *)callbackId {
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:callbackId];
}

- (void)sendError:(NSError *)error forCallbackId:(NSString *)callbackId {
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary:error.userInfo];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:callbackId];
}

#pragma mark - Consent Data conversion helper methods

- (FIRConsentType)consentTypeFromNumber:(NSNumber*)num
{
    if ([num isEqualToNumber:@1]) return FIRConsentTypeAdPersonalization;
    if ([num isEqualToNumber:@2]) return FIRConsentTypeAdStorage;
    if ([num isEqualToNumber:@3]) return FIRConsentTypeAdUserData;
    if ([num isEqualToNumber:@4]) return FIRConsentTypeAnalyticsStorage;
    return nil;
}

- (FIRConsentStatus)consentStatusFromNumber:(NSNumber *)num
{
    if ([num isEqualToNumber:@2]) return FIRConsentStatusDenied;
    if ([num isEqualToNumber:@1]) return FIRConsentStatusGranted;
    return nil;
}


@end

File: src/ios/FirebaseAnalyticsPlugin.swift
#if canImport(Cordova)
import Cordova
#endif

import AppTrackingTransparency
import FirebaseAnalytics
import FirebaseCore
import UIKit

@objc(FirebaseAnalyticsPlugin)
class FirebaseAnalyticsPlugin: CDVPlugin {
    private var manager: OSFANLManageable?

    override func pluginInitialize() {
        NSLog("Starting Firebase Analytics plugin")

        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }

        self.manager = OSFANLManagerFactory.createManager()
    }

    @objc(logEvent:)
    func logEvent(_ command: CDVInvokedUrlCommand) {
        let name = command.argument(at: 0) as? String ?? ""
        let parameters = command.argument(at: 1) as? [String: Any]

        Analytics.logEvent(name, parameters: parameters)
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(logECommerceEvent:)
    func logECommerceEvent(_ command: CDVInvokedUrlCommand) {
        guard let argumentsDictionary = command.argument(at: 0) as? [String: Any] else {
            self.sendError(OSFANLError.logEcommerceEventInputArgumentsIssue as NSError, forCallbackId: command.callbackId)
            return
        }

        do {
            let outputModel = try self.manager?.createEventModel(for: argumentsDictionary)
            guard let outputModel else {
                self.sendError(OSFANLError.logEcommerceEventInputArgumentsIssue as NSError, forCallbackId: command.callbackId)
                return
            }

            Analytics.logEvent(outputModel.name, parameters: outputModel.parameters)
            self.sendSuccessfulResult(forCallbackId: command.callbackId)
        } catch {
            self.sendError(error as NSError, forCallbackId: command.callbackId)
        }
    }

    @objc(setUserId:)
    func setUserId(_ command: CDVInvokedUrlCommand) {
        let id = command.argument(at: 0) as? String

        Analytics.setUserID(id)
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(setUserProperty:)
    func setUserProperty(_ command: CDVInvokedUrlCommand) {
        let name = command.argument(at: 0) as? String ?? ""
        let value = command.argument(at: 1) as? String

        Analytics.setUserProperty(value, forName: name)
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(setEnabled:)
    func setEnabled(_ command: CDVInvokedUrlCommand) {
        let enabled = (command.argument(at: 0) as? Bool) ?? false

        Analytics.setAnalyticsCollectionEnabled(enabled)
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(setCurrentScreen:)
    func setCurrentScreen(_ command: CDVInvokedUrlCommand) {
        let screenName = command.argument(at: 0) as? String ?? ""

        Analytics.logEvent(AnalyticsEventScreenView, parameters: [
            AnalyticsParameterScreenName: screenName
        ])
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(resetAnalyticsData:)
    func resetAnalyticsData(_ command: CDVInvokedUrlCommand) {
        Analytics.resetAnalyticsData()
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(setDefaultEventParameters:)
    func setDefaultEventParameters(_ command: CDVInvokedUrlCommand) {
        let params = command.argument(at: 0) as? [String: Any]

        Analytics.setDefaultEventParameters(params)
        self.sendSuccessfulResult(forCallbackId: command.callbackId)
    }

    @objc(requestTrackingAuthorization:)
    func requestTrackingAuthorization(_ command: CDVInvokedUrlCommand) {
        let showInformation = (command.argument(at: 0) as? Bool) ?? false

        if showInformation {
            let title = command.argument(at: 1) as? String ?? ""
            let message = command.argument(at: 2) as? String ?? ""
            let buttonTitle = command.argument(at: 3) as? String ?? ""

            self.showPermissionInformationPopup(title, message, buttonTitle) { [weak self] _ in
                self?.showTrackingAuthorizationPopup(command)
            }
        } else {
            self.showTrackingAuthorizationPopup(command)
        }
    }

    @objc(setConsent:)
    func setConsent(_ command: CDVInvokedUrlCommand) {
        do {
            let consentModel = try OSFANLConsentHelper.createConsentModel(command.arguments as NSArray)
            Analytics.setConsent(consentModel)
            self.sendSuccessfulResult(forCallbackId: command.callbackId)
        } catch {
            self.sendError(error as NSError, forCallbackId: command.callbackId)
        }
    }
}

private extension FirebaseAnalyticsPlugin {
    func showTrackingAuthorizationPopup(_ command: CDVInvokedUrlCommand) {
        if #available(iOS 14, *), Bundle.main.infoDictionary?["NSUserTrackingUsageDescription"] != nil {
            ATTrackingManager.requestTrackingAuthorization { [weak self] status in
                let result = status == .authorized
                let pluginResult = CDVPluginResult(status: .ok, messageAs: result)
                self?.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }
            return
        }

        let pluginResult = CDVPluginResult(status: .ok, messageAs: true)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    func showPermissionInformationPopup(
        _ title: String,
        _ message: String,
        _ buttonTitle: String,
        confirmationHandler: @escaping (UIAlertAction) -> Void
    ) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let okAction = UIAlertAction(title: buttonTitle, style: .default, handler: confirmationHandler)

        alert.addAction(okAction)
        self.viewController.present(alert, animated: true)
    }

    func sendSuccessfulResult(forCallbackId callbackId: String) {
        let pluginResult = CDVPluginResult(status: .ok)
        self.commandDelegate.send(pluginResult, callbackId: callbackId)
    }

    func sendError(_ error: NSError, forCallbackId callbackId: String) {
        let pluginResult = CDVPluginResult(status: .error, messageAs: error.userInfo)
        self.commandDelegate.send(pluginResult, callbackId: callbackId)
    }
}
